applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_structural_response_function.cpp

```diff
@@ -0,0 +1,752 @@
+        this->ReadDesignVariables(mConditionSensitivityScalarVariables, mConditionSensitivityVectorVariables, ResponseSettings["condition_sensitivity_variables"]);
+        
+        // Set gradient mode
+        const std::string gradient_mode = ResponseSettings["gradient_mode"].GetString();
+
+        // Mode 1: semi-analytic sensitivities
+        if (gradient_mode == "semi_analytic")
+        {
+            mGradientMode = 1;
+            double delta = ResponseSettings["step_size"].GetDouble();
+            mDelta = delta;
+        }
+        else
+            KRATOS_ERROR << "Specified gradient_mode not recognized. The only option is: semi_analytic. Specified gradient_mode: " <<  gradient_mode << std::endl;
+
+
+        KRATOS_CATCH("");
+    }
+
+    /// Destructor.
+    AdjointStructuralResponseFunction::~AdjointStructuralResponseFunction()
+    {
+    }
+
+    ModelPart& AdjointStructuralResponseFunction::GetModelPart()
+    {
+      return mrModelPart;
+    }
+
+    ModelPart& AdjointStructuralResponseFunction::GetModelPart() const 
+    {
+      return mrModelPart;
+    }
+
+    void AdjointStructuralResponseFunction::Initialize()
+    {
+        KRATOS_TRY;
+
+        this->Clear();
+        this->Check();
+
+        ModelPart& r_model_part = this->GetModelPart();
+        ModelPart& r_sensitivity_model_part = r_model_part.GetSubModelPart(mSensitivityModelPartName);
+
+        // Initialize flags.
+        VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, true, r_sensitivity_model_part.Nodes());
+        VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, true, r_sensitivity_model_part.Elements());
+        VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, true, r_sensitivity_model_part.Conditions());
+
+    if(mGradientMode == 1)
+    {
+        VariableUtils().SetNonHistoricalVariable(DISTURBANCE_MEASURE, mDelta, r_model_part.Elements());
+        VariableUtils().SetNonHistoricalVariable(DISTURBANCE_MEASURE, mDelta, r_model_part.Conditions());
+    }
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::InitializeSolutionStep(){}
+
+    void AdjointStructuralResponseFunction::FinalizeSolutionStep()
+    {
+        KRATOS_TRY;
+
+        this->UpdateSensitivities();
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::Check()
+    {
+        KRATOS_TRY;
+
+        ModelPart& r_model_part = this->GetModelPart();
+
+        KRATOS_ERROR_IF_NOT(r_model_part.HasSubModelPart(mSensitivityModelPartName))
+            << "No sub model part \"" << mSensitivityModelPartName << "\"" << std::endl;
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointStructuralResponseFunction::Clear()
+    {
+        KRATOS_TRY;
+
+        ModelPart& r_model_part = this->GetModelPart();
+
+        // Reset flags.
+        VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, false, r_model_part.Nodes());
```

Comment:
It would be better to use `Kratos::Flags` for this

Reply:
Since this is done the same way as in the adjoint fluid application, i would prefer to discuss this separately and in case change for both applications at the same time.
What is the advantage of `Kratos::Flags`?

Reply:
Kratos::Flags is faster, since it is just a Bit